File: Cpp/SortAlgorithms/MergeSort.cpp
//
//  MergeSort.cpp
//
//  Projeto e Análise de Algoritmos
//
//  Aitor Lucas, 
//  26/04/2025
//

#include "MergeSort.h"

void MergeSort::sort(std::vector<int>& arr) {
    mergeSort(arr, 0, arr.size() - 1);
}

void MergeSort::mergeSort(std::vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

void MergeSort::merge(std::vector<int>& arr, int left, int mid, int right) {
    std::vector<int> L(arr.begin() + left, arr.begin() + mid + 1);
    std::vector<int> R(arr.begin() + mid + 1, arr.begin() + right + 1);
    size_t i = 0, j = 0, k = left;
    while (i < L.size() && j < R.size()) {
        arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];
    }
    while (i < L.size()) arr[k++] = L[i++];
    while (j < R.size()) arr[k++] = R[j++];
}

File: Python/SortAlgorithms/MergeSort.py
#
#  MergeSort.py
#
#  Projeto e Análise de Algoritmos
#
#  Aitor Lucas, 
#  26/04/2025
#

from ISort import ISort

class MergeSort(ISort):
    def sort(self, arr):
        self._merge_sort(arr)

    def _merge_sort(self, arr):
        if len(arr) > 1:
            mid = len(arr) // 2
            L, R = arr[:mid], arr[mid:]
            self._merge_sort(L)
            self._merge_sort(R)

            i = j = k = 0
            while i < len(L) and j < len(R):
                arr[k] = L[i] if L[i] < R[j] else R[j]
                i += (arr[k] == L[i]) 
                j += (arr[k] == R[j]) 
                k += 1

            while i < len(L):
                arr[k] = L[i]
                i += 1
                k += 1
            while j < len(R):
                arr[k] = R[j]
                j += 1
                k += 1
